ethers-providers/src/pending_escalator.rs

```diff
@@ -0,0 +1,198 @@
+            panic!("bad args");
+        }
+
+        let first = txns.pop().expect("bad args");
+        // Sane-feeling default intervals
+        Self {
+            provider,
+            broadcast_interval: Duration::from_millis(150),
+            polling_interval: Duration::from_millis(10),
+            txns,
+            last: None,
+            sent: vec![],
+            state: PendingStates::Initial(provider.send_raw_transaction(first)),
+        }
+    }
+
+    pub fn broadcast_interval(mut self, duration: u64) -> Self {
+        self.broadcast_interval = Duration::from_secs(duration);
+        self
+    }
+
+    pub fn polling_interval(mut self, duration: u64) -> Self {
+        self.polling_interval = Duration::from_secs(duration);
+        self
+    }
+}
+
+macro_rules! check_all_receipts {
+    ($cx:ident, $this:ident) => {
+        let futs: Vec<_> = $this
+            .sent
+            .iter()
+            .map(|tx_hash| $this.provider.get_transaction_receipt(*tx_hash))
+            .collect();
+        *$this.state = CheckingReceipts(futs);
+        $cx.waker().wake_by_ref();
+        return Poll::Pending;
+    };
+}
+
+macro_rules! sleep {
+    ($cx:ident, $this:ident) => {
+        *$this.state =
+            PendingStates::Sleeping(Box::pin(tokio::time::sleep(*$this.polling_interval)));
+        $cx.waker().wake_by_ref();
+        return Poll::Pending;
+    };
+}
+
+macro_rules! completed {
+    ($this:ident, $output:expr) => {
+        *$this.state = Completed;
+        return Poll::Ready($output);
+    };
+}
+
+macro_rules! broadcast_checks {
+    ($cx:ident, $this:ident, $fut:ident) => {
+        match $fut.as_mut().poll($cx) {
+            Poll::Ready(Ok(pending)) => {
+                $this.sent.push(*pending);
+                check_all_receipts!($cx, $this);
+            }
+            Poll::Ready(Err(e)) => {
+                completed!($this, Err(e));
+            }
+            Poll::Pending => return Poll::Pending,
+        }
+    };
+}
+
+impl<'a, P> Future for EscalatingPending<'a, P>
+where
+    P: JsonRpcClient,
+{
+    type Output = Result<TransactionReceipt, ProviderError>;
+
+    fn poll(
+        self: std::pin::Pin<&mut Self>,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Self::Output> {
+        use PendingStates::*;
+
+        let this = self.project();
+
+        match this.state {
+            Initial(fut) => {
+                broadcast_checks!(cx, this, fut);
+            }
+            Sleeping(fut) => {
+                if fut.as_mut().poll(cx).is_ready() {
+                    // if timer has elapsed (or this is the first tx)
+                    if this.last.is_none()
+                        || this.last.clone().unwrap().elapsed() > *this.broadcast_interval
+                    {
+                        // then if we have a TX to broadcast, start
+                        // broadcasting it
+                        if let Some(next_to_broadcast) = this.txns.pop() {
```

Comment:
So if we have already broadcast the most escalated tx, we will just wait on that one indefinitely?

Reply:
yes